### ISSUE TITLE
Maxpool in UNet should keep the size?

### ISSUE THREAD

**Issue by DanilHu** [nie-lang/DeepImageStitching-1.0#3]
Hi, firstly thanks for opening your source code. It is really an amazing job!
Recently, I'm trying to reproduce your code in Pytorch. And I have a question about the content revision stage, specifically in UNet. I found that the function(maxpool2d in tensorflow) you use will keep the feature map size. But it seems hard to use Pytorch to do the same. Is this necessary for the result?

![kkk](https://user-images.githubusercontent.com/40864036/144693433-466f8e38-9393-4202-93b4-f34dba0c7efa.jpg)

(I've finished the Homography estimation part, and its coarse result is similar to yours. But  the result of content revision stage is pretty strange(Even when I seperately train the unet using coarse image from yours ). So I searched for your code and found that your maxpool can keep the size while mine can't). The Unet I use and the result of Unet is as follow:

![Block](https://user-images.githubusercontent.com/40864036/144693652-20f1036f-dff4-4b89-9057-f40ab04e2097.png)
![Enc_Dec](https://user-images.githubusercontent.com/40864036/144693654-1ac9c9ed-7089-4382-80c9-6663bcf288fd.png)
![Unet](https://user-images.githubusercontent.com/40864036/144693657-2e6e0119-d396-43c1-b7f9-805fdf8924d0.png)

Result
![ouput_vgg_unet](https://user-images.githubusercontent.com/40864036/144693814-d0b49a20-8689-4b69-9f89-6310161c6d5e.jpg)

**Comment by nie-lang** [comment]
Hi, I have two suggestions for your problem:

1) In my view, the padding mode ('valid' or 'same') makes no difference because the input resolution of UNET is 304 (if I remember correctly). This resolution indicates that the maxpooling operation does not pad.
2) I do not use BatchNorm in my TF implementation. I'm not sure if this difference is caused by BN.

Look forward to your TORCH implementation.

**Comment by DanilHu** [comment]
Hi, thanks very mach for replying. I've tried what you said. And surprisingly, it does work! The ghost almost disappeared.  But it comes another problem. The surrounding area becomes blurred. 

**Coarse image**


![coarse_vgg_unet20](https://user-images.githubusercontent.com/40864036/144731677-d3328166-cf84-4a6c-9426-6ca9bf9ec468.jpg)

**Unet output**
![ouput_vgg_unet20](https://user-images.githubusercontent.com/40864036/144731681-9cf39e5a-f62c-413a-a042-fb4442d1d8fc.jpg)


At first, I consider it a normal situation, since the Homography matrix is not  exactly the same as label. But I found that your result is pretty good without the blurred area. Then, I tried to change the weight of vgg loss, but it still doesn't work……So I have question as follow.
1. Why not did you use BN in your Unet? In my view, BN can make the network converge faster in CNN.
2. Did you ecounter the case of blurred area? And how you solve it?
3. what effect is the vgg loss you put in this network?(I 've tried to drop off the vgg loss, and the result is the same as using it) 
(By the way,  Deep Learning is really a metaphysics！！！-_-)

**Comment by nie-lang** [comment]
A1: There's a small bug about BN in TF, thus I avoid BN in my implementation.

A2: I do not encounter the case of blurred regions. Actually, the non-overlapping regions of the stitched images are also a little blurred (but are more clear than yours).

A3: VGG loss can improve our performance. If I remember correctly, the results will be more clear (perceptually natural) with this loss. I'm not sure about the details now, you can see the difference (image quality and PSNR/SSIM) by ablating this loss in my TF implementation.

Maybe you can check about the input of UNET (warped input1, warped input2, and coarse result) and the normalization of them (-1~1 in my implementation). Compared with the simple UNET network, the inputs and the loss functions matter more.

**Comment by dvando** [comment]
Hello @DanilHu, do you mind to share your pytorch implementation? I tried it too, but it didn't turn out the way I want it to be

**Comment by zhongchangxaing** [comment]
Hello, could you please share your corrected pytorch code, I would be very grateful！ @DanilHu
